src/app/games/hive/HiveRules.ts

```diff
@@ -0,0 +1,237 @@
+        const sourcePieceStack: HivePieceStack = state.getAt(move.coord);
+        const destinationPieceStack: HivePieceStack = state.getAt(move.end);
+        newState.setAt(move.coord, sourcePieceStack.removeTopPiece());
+        newState.setAt(move.end, destinationPieceStack.add(sourcePieceStack.topPiece()));
+        return newState.increaseTurnAndRecomputeBounds();
+    }
+
+    public isLegal(move: HiveMove, state: HiveState): MGPFallible<void> {
+        if (move instanceof HiveMoveDrop) {
+            return this.isLegalDrop(move, state);
+        } else if (move instanceof HiveMoveCoordToCoord) {
+            return this.isLegalMoveCoordToCoord(move, state);
+        } else {
+            // Move is pass
+            if (this.shouldPass(state)) {
+                return MGPFallible.success(undefined);
+            } else {
+                return MGPFallible.failure(RulesFailure.CANNOT_PASS());
+            }
+        }
+    }
+
+    public isLegalMoveCoordToCoord(move: HiveMoveCoordToCoord, state: HiveState): MGPFallible<void> {
+        if (state.queenBeeLocation(state.getCurrentPlayer()).isPresent() === false) {
+            return MGPFallible.failure(HiveFailure.QUEEN_BEE_MUST_BE_ON_BOARD_BEFORE_MOVE());
+        }
+        const stack: HivePieceStack = state.getAt(move.coord);
+        if (stack.isEmpty()) {
+            return MGPFallible.failure(RulesFailure.MUST_CHOOSE_PLAYER_PIECE());
+        }
+        const movedPiece: HivePiece = stack.topPiece();
+        if (movedPiece.owner !== state.getCurrentPlayer()) {
+            return MGPFallible.failure(RulesFailure.MUST_CHOOSE_PLAYER_PIECE());
+        }
+
+        const moveValidity: MGPFallible<void> = HivePieceBehaviour.from(movedPiece).moveValidity(move, state);
+        if (moveValidity.isFailure()) {
+            return moveValidity;
+        }
+
+        const stateWithoutMovedPiece: HiveState = state.getCopy();
+        stateWithoutMovedPiece.setAt(move.coord, stack.removeTopPiece());
+        if (stateWithoutMovedPiece.isDisconnected()) {
+            // Cannot disconnect the hive, even in the middle of a move
+            return MGPFallible.failure(HiveFailure.CANNOT_DISCONNECT_HIVE());
+        }
+
+        if (stateWithoutMovedPiece.numberOfNeighbors(move.end) === 0) {
+            // Cannot disconnect the hive after the move, i.e., the destination should have one occupied neighbor
+            return MGPFallible.failure(HiveFailure.CANNOT_DISCONNECT_HIVE());
+        }
+
+        return MGPFallible.success(undefined);
+    }
+
+    public mustPlaceQueenBee(state: HiveState): boolean {
+        return state.turn >= 6 && state.queenBeeLocation(state.getCurrentPlayer()).isAbsent();
+    }
+
+    public isLegalDrop(move: HiveMoveDrop, state: HiveState): MGPFallible<void> {
+        const player: Player = state.getCurrentPlayer();
+
+        // This should be a piece of the player
+        if (move.piece.owner !== state.getCurrentPlayer()) {
```

Comment:
prefer positive comparison !

Reply:
pourquoi ? C'est littéralement `if (conditiond'Erreur) { erreur }` avec condition d'erreur = la pièce n'appartient pas au joueur

Reply:
bah, owner est un Player, pas un PlayerOrNone, donc why the negation, "si tu bouge l'adversaire" = condition d'Erreur.

Là t'as fait if (condition de success === false) { erreur } !